### PR TITLE
Adding TCP Info header support to header rewrite

### PR DESCRIPTION
**Pull request by dbhat**
Opening new pull request. This is squashed commit of closed PR #7441.

### CODE REVIEW

**Comment by dbhat** [comment]
Requesting @bryancall for review

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by maskit** [comment]
header_rewrite_l_value test failed once on the previous PR as well. It may happen intermittently.

**Comment by ezelkow1** [comment]
Since I wrote that lvalue test, I wonder if this is a curl difference popping up maybe? The test is failure simple to just check for a regression where the `[L]` end of a header rewrite rule was being leaked back into the header it was writing. So thats all its doing is adding a header and making sure `[L]` isnt in there

**Comment by ezelkow1** [comment]
if you check the autest, theres actually a fair amount of tests that fail all around there, not just lvalue, but this is that one tests output
` Reason: File differences
           Gold File : /var/jenkins/workspace/autest-github/src/tests/gold_tests/pluginTest/header_rewrite/gold/header_rewrite-l_value.gold
           Data File : /var/tmp/ausb-7516.14844/header_rewrite_l_value/_output/0-tr-Default/stream.stderr.txt
             *   Trying 127.0.0.1:61289...
             * TCP_NODELAY set
               % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                              Dload  Upload   Total   Spent    Left  Speed
             
               0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 61289 (#0)
             > GET http://www.example.com/ HTTP/1.1
             > Host: www.example.com
             > User-Agent: curl/7.65.0-DEV
             > Accept: */*
             > Proxy-Connection: Keep-Alive
             > Proxy-Connection: keep-alive
             > 
             * Mark bundle as not supporting multiuse
             < HTTP/1.1 200 OK
             < Date: Thu, 11 Feb 2021 23:10:39 GMT
             < Age: 0
             < Transfer-Encoding: chunked
             < Proxy-Connection: keep-alive
           - < Server: ATS/``
           ?               ^^
           
           + < Server: ATS/10.0.0
           ?               ^^^^^^
           
           - < X-First: First
           - < X-Last: Last
             < 
             { [5 bytes data]
             
               0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
             * Connection #0 to host 127.0.0.1 left intact`

File: plugins/header_rewrite/conditions.cc
Comment (zwoop) [inline plugins/header_rewrite/conditions.cc:1399]: Something like this is probably more efficient:

```
    char buf[12 * 4 + 5]; // 4 uint32's

    snprintf(buf, sizeof(buf), "%u; %u; %u; %u; ", rtt, rto, snd_cwnd, all_rtx);

    s += buf;
```

Although, use PRIu32 in the printf.

File: plugins/header_rewrite/conditions.cc
Comment (zwoop) [inline plugins/header_rewrite/conditions.cc:1390]: why do you need these intermediary / temp variables? Just use info.tcp_rtt instead, it's much clearer, and less waste.

File: plugins/header_rewrite/factory.cc
Comment (zwoop) [inline plugins/header_rewrite/factory.cc:142]: For consistency with other conditions, I think a name like

```
TCP-INFO
```

would be preferable. But I don't feel strongly.

File: plugins/header_rewrite/parser.cc
Comment (zwoop) [inline plugins/header_rewrite/parser.cc:247]: I don't remember how we deal with things like this, but this hook would not be viable when invoked as a remap plugin. TXN_START happens before that; meaning, this can only be used when running header_rewrite as a global plugin.

File: doc/admin-guide/plugins/header_rewrite.en.rst
Comment (zwoop) [inline doc/admin-guide/plugins/header_rewrite.en.rst:949]: At a minimum, I think we need to document that this would only work as a global plugin.

File: doc/admin-guide/plugins/header_rewrite.en.rst
Comment (zwoop) [inline doc/admin-guide/plugins/header_rewrite.en.rst:498]: this looks weird, the syntax would be, I assume

```
add-header @PropertyName "%{TCPIINFO}"
```

no ?

File: plugins/header_rewrite/conditions.cc
Comment (zwoop) [inline plugins/header_rewrite/conditions.cc:1374]: I wonder in the case of there being no TCP_INFO available on the platform (for whatever reason, seems odd on its own), should it not at least set something here, such as "-" ?

Alternatively, since we have eliminated Solaris, and only support macOS, Linux and FreeBSD, maybe we don't need this check at all, and assume that TCP_INFO is available ?

File: plugins/header_rewrite/conditions.cc
Comment (zwoop) [inline plugins/header_rewrite/conditions.cc:1336]: Is there a missing set of #include's here somewhere ? Don't you need something similar to the include jive from the old tcpinfo plugin?

```
// This is a bit of a hack, to get the more linux specific tcp_info struct ...
#if HAVE_STRUCT_LINUX_TCP_INFO
#include <linux/tcp.h>
#elif HAVE_NETINET_IN_H
#include <netinet/tcp.h>
#endif
```

Of course, if you don't need the linux specific ones, only include the netinet/tcp.h include, but assuming this even compiles, something else is dragging in tcp.h? But depending on that is not good. It seems lulu.h includes netinet/in.h, so maybe add a netinet/tcp.h there too ?

**Review by zwoop — CHANGES_REQUESTED** [review]
See comments, looks good, but needs a few minor tweaks.

**Comment by bryancall** [comment]
[approve ci autest]

File: plugins/header_rewrite/parser.cc
Comment (dbhat) [reply plugins/header_rewrite/parser.cc:247]: If TXN_START is invoked in remap, the header value is not added.

File: plugins/header_rewrite/conditions.cc
Comment (dbhat) [reply plugins/header_rewrite/conditions.cc:1385]: Decided to keep the check and insert "-" when TCP_INFO is not available

**Comment by dbhat** [comment]
@zwoop: Done with changes, requesting review

File: plugins/header_rewrite/parser.cc
Comment (zwoop) [reply plugins/header_rewrite/parser.cc:247]: Yeh, but I'm thinking it should maybe be a warning or even an error if TXN_START_HOOK is used in a remap plugin.

Comment (dbhat) [reply plugins/header_rewrite/parser.cc:247]: I believe a warning would suffice.

File: plugins/header_rewrite/resources.cc
Comment (maskit) [inline plugins/header_rewrite/resources.cc:90]: I'm not familiar with this area but this block doesn't seem to be correct. If I read the code correctly, `bufp` and `hdr_loc` point to headers that will be modified on current event (hook). Since this block is under `case TS_HTTP_SEND_RESPONSE_HDR_HOOK:`, those should point to response headers.

`TSHttpTxnClientRespGet` sets the right pointers but this block overwrites those with the ones for client request headers.

Removing this block fixes header_rewrite_l_value and header_rewrite_cond_ssn_txn_count at minimum.

File: plugins/header_rewrite/resources.cc
Comment (maskit) [inline plugins/header_rewrite/resources.cc:109]: If I'm reading the code correctly, this condition would never be true because request header has not been read on `TS_HTTP_TXN_START_HOOK` yet.
https://docs.trafficserver.apache.org/en/9.0.x/developer-guide/plugins/hooks-and-transactions/index.en.html

Comment (maskit) [reply plugins/header_rewrite/resources.cc:109]: Actually, the pointers are available. We can't read request headers, but can add headers.

File: plugins/header_rewrite/resources.cc
Comment (maskit) [inline plugins/header_rewrite/resources.cc:120]: I'd expect headers added on TS_HTTP_TXN_CLOSE_HOOK are available in client *response* header. This code seems like to add headers to client *request* headers.

**Comment by maskit** [comment]
[approve ci autest]

**Review by maskit — DISMISSED** [review]
Looks good to me.

It would be more user friendly if there was an explanation about which header (e.g. client request header, client response header, etc) will be modified on each hook (especially for TXN_CLOSE). It's easy to guess to me, but maybe not for new users.

**Review by zwoop — APPROVED** [review]
Looks good, but would like one more non-Apple reviewer before we merge.

Thanks for addressing all the comments!

**Review by sudheerv — COMMENTED** [review]
I'm also curious on how this information is intended to be used. If the intention is for mainly exposing the TCP-INFO to access logs, I wonder if it makes sense to consider using log tags instead of header-rewrite operator?

**Comment by zwoop** [comment]
> I'm also curious on how this information is intended to be used. If the intention is for mainly exposing the TCP-INFO to access logs, I wonder if it makes sense to consider using log tags instead of header-rewrite operator?

Yeh, primarily for logs. Moving it to the core is an option, but the thought was to avoid polluting the core more with a bunch of #ifdef's (and also makes it more difficult for people to make custom additions or changes). But if the consensus is that it'd be better to move this to the core, rather than a plugin, I'm fine with that too.

File: plugins/header_rewrite/conditions.cc
Comment (dbhat) [reply plugins/header_rewrite/conditions.cc:1395]: Implemented (a) as follows:

Comment (sudheerv) [reply plugins/header_rewrite/conditions.cc:1395]: Looks good. A minor nit though is to change it to just ```if (TSHttpTxnIsInternal(res.txnp)) {``` (== 1 should be implied).

File: doc/admin-guide/plugins/header_rewrite.en.rst
Comment (sudheerv) [inline doc/admin-guide/plugins/header_rewrite.en.rst:502]: Perhaps, reword it as 

"This operation is not supported on transactions originated within Traffic Server (for e.g using the TS API TSHttpTxnIsInternal()  with a ref to the API doc link)".

File: plugins/header_rewrite/conditions.cc
Comment (dbhat) [reply plugins/header_rewrite/conditions.cc:1395]: Changed.

File: doc/admin-guide/plugins/header_rewrite.en.rst
Comment (dbhat) [reply doc/admin-guide/plugins/header_rewrite.en.rst:502]: Reworded to clarify.

**Review by sudheerv — APPROVED** [review]
LGTM!! 

Awesome job and thank you for taking care of the review comments. 

Please make sure to squash the commits before merging :)

**Comment by maskit** [comment]
CI jobs were not started somehow. Trying to start them manually. [approve ci]

Also, squashing commits is no longer required on author's side. It will be done automatically when a committer click the merge button.

**Comment by maskit** [comment]
[approve ci]

**Comment by sudheerv** [comment]
> CI jobs were not started somehow. Trying to start them manually. [approve ci]
> 
> Also, squashing commits is no longer required on author's side. It will be done automatically when a committer click the merge button.

Hmm..I thought it defaults to last used option and not always ensure squash and commit option during merge. Assuming that is the case, I’d always remind the author to squash before merging (doesn’t really matter if it’s done manually or using the GitHub UI option)

**Comment by maskit** [comment]
> I thought it defaults to last used option 

That's true, however, we disabled the other options on this reporisotry. So we can't merge PRs without squashing commits.

image

**Review by maskit — APPROVED** [review]

**Comment by zwoop** [comment]
Cherry-picked to v9.1.x branch.

**Pull request merged**
